<?xml version="1.0"?>
<robot name="custom_arm">

    <link name="world"/>

<!-- ......................Base Link & Joint 0............................................................ -->

    <link name="base_link">
        <visual>
            <origin xyz="0 0 0.025" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.05" radius="0.1"/>
            </geometry>
            <material name="silver">
                <color rgba="0.75 0.75 0.75 1"/>
            </material>
        </visual>

        <collision>
            <origin xyz="0 0 0.025" rpy="0 0 0" />
            <geometry>
                <cylinder length="0.05" radius="0.1"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0.025" rpy="0 0 0"/>
            <mass value="5.0"/>
            <inertia ixx="0.0135" ixy="0.0" ixz="0.0" iyy="0.0135" iyz="0.0" izz="0.05"/>
        </inertial>
    </link>

    <joint name="joint_0" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
    </joint>

<!-- ......................Link 1 & Joint 1............................................................ -->

    <link name="link_1">
        <visual>
            <origin xyz="0 0 0.25" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.5" radius="0.05"/>
            </geometry>
            <material name="silver">
                <color rgba="0.75 0.75 0.75 1"/>
            </material>
        </visual>

        <collision>
            <origin xyz="0 0 0.25" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.5" radius="0.05"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0.25" rpy="0 0 0"/>
            <mass value="5.0"/>
            <inertia ixx="0.107" ixy="0.0" ixz="0.0" iyy="0.107" iyz="0.0" izz="0.0125"/>
        </inertial>
    </link>
  
    <joint name="joint_1" type="continuous">
        <axis xyz="0 0 1"/> <!-- Child link will rotate on this axis -->
        <parent link="base_link"/>
        <child link="link_1"/>
        <origin xyz="0 0 0.05" rpy="0 0 0"/> <!-- Parent link position and orientation is 0 in 3 axes. 
                                                    Allocate child link by giving desired position and orietation data -->
    </joint>

<!-- .................Link 2 & Joint 2.................................................... -->

    <link name="link_2">
        <visual>
            <origin xyz="0 0 0.2" rpy="0 0 0"/> <!-- Where should be the center(axes) of link -->
            <geometry>
                <cylinder length="0.4" radius="0.05"/>
            </geometry>
            <material name="silver"/>
        </visual>

        <collision>
            <origin xyz="0 0 0.2" rpy="0 0 0"/> <!-- repeat visual -->
            <geometry>
                <cylinder length="0.4" radius="0.05"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
            <mass value="2.0"/>
            <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
        </inertial>
    </link>

    <joint name="joint_2" type="continuous">
        <axis xyz="0 1 0"/>
        <parent link="link_1"/>
        <child link="link_2"/>
        <origin xyz="0 -0.1 0.45" rpy="0 1.5708 0"/> 
    </joint>

<!-- ............Link 3 & Joint 3..................................................... -->

    <link name="link_3">        
        <visual>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.4" radius="0.05"/>
            </geometry>
            <material name="silver"/>
            </visual>

        <collision>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.4" radius="0.05"/>
            </geometry>
        </collision>

        <inertial>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
            <mass value="2.0"/>
            <inertia ixx="0.027" ixy="0.0" ixz="0.0" iyy="0.027" iyz="0.0" izz="0.0025"/>
        </inertial>
    </link>

    <joint name="joint_3" type="continuous">
        <axis xyz="0 1 0"/>
        <parent link="link_2"/>
        <child link="link_3"/>
        <origin xyz="0.0 0.1 0.35" rpy="0 0 0"/>
    </joint>

<!-- .................Link 4 & Joint 4................................................... -->

    <link name="link_4">
        <visual>
            <geometry>
                <box size="0.05 0.05 0.05"/>
            </geometry>
            <material name="silver"/>
        </visual>

        <collision>
            <geometry>
                <box size="0.05 0.05 0.05"/>
            </geometry>
        </collision>
        
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.00042" iyy="0.00042" izz="0.00042" ixy="0" ixz="0" iyz="0"/>
        </inertial>    
    </link>

    <joint name="joint_4" type="continuous">
        <axis xyz="0 1 0"/>
        <parent link="link_3"/>
        <child link="link_4"/>
        <origin xyz="0.0 0.0 0.425" rpy="0 0 0"/>
    </joint> 

<!-- ...................Colours......................................... -->

    <gazebo reference="base_link">
        <material>Gazebo/Yellow</material>
    </gazebo>
    
    <gazebo reference="link_1">
        <material>Gazebo/Blue</material>
    </gazebo>
    
    <gazebo reference="link_2">
        <material>Gazebo/Yellow</material>
    </gazebo>
    
    <gazebo reference="link_3">
        <material>Gazebo/Blue</material>
    </gazebo>
    
    <gazebo reference="link_4">
        <material>Gazebo/Yellow</material>
    </gazebo>

<!-- ...................Transmission......................................... -->

    <transmission name ="trans_1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_1">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_1">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name ="trans_2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_2">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_2">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name ="trans_3">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_3">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_3">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name ="trans_4">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_4">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_4">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

<!-- ...................Controller......................................... -->

    <gazebo>
        <plugin name="control" filename="libgazebo_ros_control.so"/>
    </gazebo>

</robot>